### ISSUE TITLE
Docker File not working for Arm v8

### ISSUE THREAD

**Issue by sudipmandal** [Budibase/budibase#5331]
**Describe the bug**
The docker file in this repo located at https://github.com/Budibase/budibase/blob/develop/hosting/single/Dockerfile does not seem to work for 64 bit ARM v8 Arch machines.
As per this merge https://github.com/Budibase/budibase/pull/5209 the dockerfile should (theoritically) work for Arm v8

I have used docker buildx to build the dockerfile as I could not find any official armv8 image on docker hub, the dockerfile builds successfully however when trying to run this on a 64bit Armv8 device (ROC-RK3328-CC chipset) the container fails to start, throws garbage data when connected to container console. (see screenshot below)

**To Reproduce**
1. On any Arm v8 device with docker installed, build the single/Dockerfile (alternatively pull the sudipthegreat/budibase-arm:v1.0.110 image)
2. Start the Container using docker run, in my case did something like this
`docker run --name budibase \`
   ` --publish 10000:10000 \`
    `--volume /path/to/data-persistence:/opt/couchdb/data \`
    `--volume /path/to/minio-persistence:/minio \`
    `--network app-tier \`
    `--env COUCHDB_PASSWORD=**** \`
    `--env COUCHDB_USER=budibase \`
    `--env COUCH_DB_URL=http://budibase:***@localhost:5984 \`
    `--env BUDIBASE_ENVIRONMENT=PRODUCTION \`
    `--env MINIO_URL=http://localhost:9000 \`
    `--env REDIS_URL=localhost:6379 \`
    `--env WORKER_URL=http://localhost:4002 \`
    `--env INTERNAL_API_KEY=YOURAPIKEY \`
    `--env JWT_SECRET=YOURJWTSECRET \`
    `--env MINIO_ACCESS_KEY=budibase \`
    `--env MINIO_SECRET_KEY=budibase \`
    `--env SELF_HOSTED=1 \`
    `--env CLUSTER_PORT=10000 \`
    `--env REDIS_PASSWORD=budibase \`
    `--env ARCHITECTURE=arm \`
    `--env APP_PORT=4001 \`
    `--env WORKER_PORT=4002 \`
    `--env NODE_ENV=production \`
    `sudipthegreat/budibase-arm:v1.0.110 `

Pretty standard docker run command, only change done was to the env "Architecture=arm" in the dockerfile this is set as "amd".

**Expected behavior**
Container should run and budibase should open up on http://serverip:10000

**Screenshots**
the docker container console shows this
![image](https://user-images.githubusercontent.com/3803018/162553521-bbda0d05-d645-4e12-b99e-7328d8dec696.png)

**Desktop (please complete the following information):**
 - OS: Armbian Linux 5.15
 - Device : Libre Renegade Firefly (https://en.t-firefly.com/product/rocrk3328cc.html)
 - Browser: N/A (budibase fails to start)
 - Version : Budibase develop branch 

**Additional context**

**Comment by sudipmandal** [comment]
For reference this is how I have built the Armv8 image and pushed to docker hub

cd $(Build.SourcesDirectory)

docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
docker buildx create --name multiarch --driver docker-container --use
docker buildx inspect --bootstrap

docker buildx build -f hosting/single/Dockerfile -t sudipthegreat/budibase-arm:v1.0.110 --platform linux/arm64 --push .

**Comment by sudipmandal** [comment]
**Update** : Seems the dockerfile was not fully ARM compliant, and there was x64 code being copied from the repo which was causing the above issue.

Had to make the changes listed below to make the dockerfile work and get the docker image build correctly for ARM v8 architecture, budibase still does not load but a big improvement is that now the console shows meaningful info and browser loads on port 10000 but shows 502 bad gateway error.

_________________
**Changes made to the repo to make docker file build correctly (hopefully) for ARM v8**
Clone repo and make following changes before building with buildx

**Modify /hosting)/single/dockerfile**
Change Line 17 to : `ENV ARCHITECTURE=arm`

**Modify packages/server/scripts/integrations/oracle/instantclient/linux/x86-64/install.sh**
Remove Line 13
Add the following line After Line 14 `wget -O basiclite-21.4.zip https://download.oracle.com/otn_software/linux/instantclient/191000/instantclient-basiclite-linux.arm64-19.10.0.0.0dbru.zip`

Run the docker buildx command above.
___________________

I know this is not the best way to do this, but its what I could do due to time constraints, ideally the oracle instantclient install.sh script should either detect the architecture or read it as a parameter and download the correct installclient zip file from the web

Looking into the console log to see what might be causing the 502 error, if anyone has any ideas, please share your thoughts, it would help a lot of home users run this on their raspberry pis and arm devices for making personal apps for home use.





**Console Log:**
└─────┴───────────┴─────────────┴─────────┴─────────┴──────────┴────────┴──────┴───────────┴──────────┴──────────┴──────────┴──────────┘
/
[notice] 2022-04-09T08:08:06.869856Z couchdb@127.0.0.1 <0.371.0> -------- chttpd_auth_cache changes listener died because the _users database does not exist. Create the database to silence this notice.
[error] 2022-04-09T08:08:06.872018Z couchdb@127.0.0.1 emulator -------- Error in process <0.521.0> on node 'couchdb@127.0.0.1' with exit value:
{database_does_not_exist,[{mem3_shards,load_shards_from_db,"_users",[{file,"src/mem3_shards.erl"},{line,400}]},{mem3_shards,load_shards_from_disk,1,[{file,"src/mem3_shards.erl"},{line,375}]},{mem3_shards,load_shards_from_disk,2,[{file,"src/mem3_shards.erl"},{line,404}]},{mem3_shards,for_docid,3,[{file,"src/mem3_shards.erl"},{line,97}]},{fabric_doc_open,go,3,[{file,"src/fabric_doc_open.erl"},{line,39}]},{chttpd_auth_cache,ensure_auth_ddoc_exists,2,[{file,"src/chttpd_auth_cache.erl"},{line,198}]},{chttpd_auth_cache,listen_for_changes,1,[{file,"src/chttpd_auth_cache.erl"},{line,145}]}]}

[error] 2022-04-09T08:08:06.872813Z couchdb@127.0.0.1 emulator -------- Error in process <0.521.0> on node 'couchdb@127.0.0.1' with exit value:
{database_does_not_exist,[{mem3_shards,load_shards_from_db,"_users",[{file,"src/mem3_shards.erl"},{line,400}]},{mem3_shards,load_shards_from_disk,1,[{file,"src/mem3_shards.erl"},{line,375}]},{mem3_shards,load_shards_from_disk,2,[{file,"src/mem3_shards.erl"},{line,404}]},{mem3_shards,for_docid,3,[{file,"src/mem3_shards.erl"},{line,97}]},{fabric_doc_open,go,3,[{file,"src/fabric_doc_open.erl"},{line,39}]},{chttpd_auth_cache,ensure_auth_ddoc_exists,2,[{file,"src/chttpd_auth_cache.erl"},{line,198}]},{chttpd_auth_cache,listen_for_changes,1,[{file,"src/chttpd_auth_cache.erl"},{line,145}]}]}

[notice] 2022-04-09T08:08:06.915367Z couchdb@127.0.0.1 <0.393.0> -------- couch_replicator_clustering : cluster stable
[notice] 2022-04-09T08:08:06.932924Z couchdb@127.0.0.1 <0.403.0> -------- Started replicator db changes listener <0.522.0>
[info] 2022-04-09T08:08:06.937286Z couchdb@127.0.0.1 <0.524.0> -------- open_result error {not_found,no_db_file} for _replicator
[notice] 2022-04-09T08:08:11.872510Z couchdb@127.0.0.1 <0.371.0> -------- chttpd_auth_cache changes listener died because the _users database does not exist. Create the database to silence this notice.
[error] 2022-04-09T08:08:11.874316Z couchdb@127.0.0.1 emulator -------- Error in process <0.576.0> on node 'couchdb@127.0.0.1' with exit value:
{database_does_not_exist,[{mem3_shards,load_shards_from_db,"_users",[{file,"src/mem3_shards.erl"},{line,400}]},{mem3_shards,load_shards_from_disk,1,[{file,"src/mem3_shards.erl"},{line,375}]},{mem3_shards,load_shards_from_disk,2,[{file,"src/mem3_shards.erl"},{line,404}]},{mem3_shards,for_docid,3,[{file,"src/mem3_shards.erl"},{line,97}]},{fabric_doc_open,go,3,[{file,"src/fabric_doc_open.erl"},{line,39}]},{chttpd_auth_cache,ensure_auth_ddoc_exists,2,[{file,"src/chttpd_auth_cache.erl"},{line,198}]},{chttpd_auth_cache,listen_for_changes,1,[{file,"src/chttpd_auth_cache.erl"},{line,145}]}]}

[error] 2022-04-09T08:08:11.875406Z couchdb@127.0.0.1 emulator -------- Error in process <0.576.0> on node 'couchdb@127.0.0.1' with exit value:
{database_does_not_exist,[{mem3_shards,load_shards_from_db,"_users",[{file,"src/mem3_shards.erl"},{line,400}]},{mem3_shards,load_shards_from_disk,1,[{file,"src/mem3_shards.erl"},{line,375}]},{mem3_shards,load_shards_from_disk,2,[{file,"src/mem3_shards.erl"},{line,404}]},{mem3_shards,for_docid,3,[{file,"src/mem3_shards.erl"},{line,97}]},{fabric_doc_open,go,3,[{file,"src/fabric_doc_open.erl"},{line,39}]},{chttpd_auth_cache,ensure_auth_ddoc_exists,2,[{file,"src/chttpd_auth_cache.erl"},{line,198}]},{chttpd_auth_cache,listen_for_changes,1,[{file,"src/chttpd_auth_cache.erl"},{line,145}]}]}

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0[error] 2022-04-09T08:08:13.923375Z couchdb@127.0.0.1 <0.577.0> 26fac2052c Request to create N=3 DB but only 1 node(s)

100    12  100    12    0     0     27      0 --:--:-- --:--:-- --:--:--    27
100    12  100    12    0     0     27      0 --:--:-- --:--:-- --:--:--    27
{"ok":true}
[notice] 2022-04-09T08:08:14.190323Z couchdb@127.0.0.1 <0.577.0> 26fac2052c localhost:5984 127.0.0.1 sudip PUT /_users 201 ok 400
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0[error] 2022-04-09T08:08:14.248338Z couchdb@127.0.0.1 <0.578.0> a049dd7960 Request to create N=3 DB but only 1 node(s)

100    [notice] 2022-04-09T08:08:14.273347Z couchdb@127.0.0.1 <0.578.0> a049dd7960 localhost:5984 127.0.0.1 sudip PUT /_replicator 201 ok 29
12  100    12    0     0    333      0 --:--:-- --:--:-- --:--:--   333
{"ok":true}
[notice] 2022-04-09T08:08:14.287056Z couchdb@127.0.0.1 <0.522.0> -------- creating replicator ddoc <<"shards/80000000-ffffffff/_replicator.1649491694">>
[notice] 2022-04-09T08:08:16.002231Z couchdb@127.0.0.1 <0.297.0> -------- rexi_server : cluster stable
[notice] 2022-04-09T08:08:16.007251Z couchdb@127.0.0.1 <0.301.0> -------- rexi_buffer : cluster stable
/runner.sh: line 3: /minio/minio: No such file or directory
Restarting nginx: nginx9:C 09 Apr 2022 08:12:19.116 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
9:C 09 Apr 2022 08:12:19.117 # Redis version=6.0.16, bits=64, commit=00000000, modified=0, pid=9, just started
9:C 09 Apr 2022 08:12:19.117 # Configuration loaded
9:M 09 Apr 2022 08:12:19.128 * Running mode=standalone, port=6379.
9:M 09 Apr 2022 08:12:19.128 # Server initialized
9:M 09 Apr 2022 08:12:19.129 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
9:M 09 Apr 2022 08:12:19.145 * Ready to accept connections
.
/app /
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[PM2] Spawning PM2 daemon with pm2_home=/root/.pm2
[info] 2022-04-09T08:12:23.371072Z couchdb@127.0.0.1 <0.255.0> -------- Preflight check: Asserting Admin Account

[info] 2022-04-09T08:12:23.386331Z couchdb@127.0.0.1 <0.255.0> -------- Apache CouchDB 3.2.1 is starting.

[info] 2022-04-09T08:12:23.386613Z couchdb@127.0.0.1 <0.256.0> -------- Starting couch_sup
[PM2] PM2 Successfully daemonized
[PM2] Starting /bin/bash in fork_mode (1 instance)
[info] 2022-04-09T08:12:23.725823Z couchdb@127.0.0.1 <0.255.0> -------- Apache CouchDB has started. Time to relax.

[notice] 2022-04-09T08:12:23.784931Z couchdb@127.0.0.1 <0.296.0> -------- rexi_server : started servers
[notice] 2022-04-09T08:12:23.795856Z couchdb@127.0.0.1 <0.300.0> -------- rexi_buffer : started servers
[PM2] Done.
┌─────┬────────┬─────────────┬─────────┬─────────┬──────────┬────────┬──────┬───────────┬──────────┬──────────┬──────────┬──────────┐
│ id  │ name   │ namespace   │ version │ mode    │ pid      │ uptime │ ↺    │ status    │ cpu      │ mem      │ user     │ watching │
├─────┼────────┼─────────────┼─────────┼─────────┼──────────┼────────┼──────┼───────────┼──────────┼──────────┼──────────┼──────────┤
│ 0   │ app    │ default     │ N/A     │ fork    │ 149      │ 0s     │ 0    │ online    │ 0%       │ 26.2mb   │ root     │ disabled │
└─────┴────────┴─────────────┴─────────┴─────────┴──────────┴────────┴──────┴───────────┴──────────┴──────────┴──────────┴──────────┘
/

/
[notice] 2022-04-09T08:12:24.161541Z couchdb@127.0.0.1 <0.330.0> -------- mem3_reshard_dbdoc start init()
[notice] 2022-04-09T08:12:24.180080Z couchdb@127.0.0.1 <0.332.0> -------- mem3_reshard start init()
[notice] 2022-04-09T08:12:24.180641Z couchdb@127.0.0.1 <0.333.0> -------- mem3_reshard db monitor <0.333.0> starting
[notice] 2022-04-09T08:12:24.190794Z couchdb@127.0.0.1 <0.332.0> -------- mem3_reshard starting reloading jobs
[notice] 2022-04-09T08:12:24.191891Z couchdb@127.0.0.1 <0.332.0> -------- mem3_reshard finished reloading jobs
[info] 2022-04-09T08:12:24.323642Z couchdb@127.0.0.1 <0.339.0> -------- Apache CouchDB has started. Time to relax.

[info] 2022-04-09T08:12:24.324045Z couchdb@127.0.0.1 <0.339.0> -------- Apache CouchDB has started on http://any:5984/
[notice] 2022-04-09T08:12:24.795615Z couchdb@127.0.0.1 <0.429.0> -------- All system databases exist.
[PM2] Starting /bin/bash in fork_mode (1 instance)
[PM2] Done.
┌─────┬───────────┬─────────────┬─────────┬─────────┬──────────┬────────┬──────┬───────────┬──────────┬──────────┬──────────┬──────────┐
│ id  │ name      │ namespace   │ version │ mode    │ pid      │ uptime │ ↺    │ status    │ cpu      │ mem      │ user     │ watching │
├─────┼───────────┼─────────────┼─────────┼─────────┼──────────┼────────┼──────┼───────────┼──────────┼──────────┼──────────┼──────────┤
│ 0   │ app       │ default     │ N/A     │ fork    │ 149      │ 1s     │ 0    │ online    │ 22.2%    │ 76.2mb   │ root     │ disabled │
│ 1   │ worker    │ default     │ N/A     │ fork    │ 178      │ 0s     │ 0    │ online    │ 0%       │ 17.9mb   │ root     │ disabled │
└─────┴───────────┴─────────────┴─────────┴─────────┴──────────┴────────┴──────┴───────────┴──────────┴──────────┴──────────┴──────────┘
/
[notice] 2022-04-09T08:12:29.431156Z couchdb@127.0.0.1 <0.386.0> -------- couch_replicator_clustering : cluster stable
[notice] 2022-04-09T08:12:29.443382Z couchdb@127.0.0.1 <0.413.0> -------- Started replicator db changes listener <0.531.0>
[info] 2022-04-09T08:12:29.446625Z couchdb@127.0.0.1 <0.533.0> -------- open_result error {not_found,no_db_file} for _replicator
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100    95  100    95    0     0    317      0 --:--:-- --:--:-- --:--:--   318
{"error":"file_exists","reason":"The database could not be created, the file already exists."}
[notice] 2022-04-09T08:12:36.032943Z couchdb@127.0.0.1 <0.580.0> f589e46459 localhost:5984 127.0.0.1 sudip PUT /_users 412 ok 267
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0[notice] 2022-04-09T08:12:36.101271Z couchdb@127.0.0.1 <0.581.0> 1d95c5f13e localhost:5984 127.0.0.1 sudip PUT /_replicator 412 ok 4
{"error":"file_exists","reason":"The database could not be created, the file already exists."}

100    95  100    95    0     0   3518      0 --:--:-- --:--:-- --:--:--  3518
[notice] 2022-04-09T08:12:38.785279Z couchdb@127.0.0.1 <0.296.0> -------- rexi_server : cluster stable
[notice] 2022-04-09T08:12:38.788089Z couchdb@127.0.0.1 <0.300.0> -------- rexi_buffer : cluster stable

**Comment by mike12345567** [comment]
Hi @sudipmandal - this is not a released feature yet - one of the goals for the single container deployment is to support ARM architectures (as our docker compose setup does not) but so far there has only been a very small amount of research work put into this, once issue #5202 is complete and we start publishing images then we will be offering full, proper support for this deployment method, but for now it is still very much a work in progress - I would not trust it as a deployment target right now as it is in such early days still.

**Comment by sudipmandal** [comment]
@mike12345567 Oh ok no worries, looks like I jumped the gun on this :)
Thank you for all the work, super excited and look forward to the official ARM images once available.

If it's of any help, I looked into the build and node logs, the error is because ARM does not offer pre-built images for node-oracledb, this module is marked as optional, but for whatever reason the app node keeps restarting with this error.
I found this discussion https://github.com/oracle/node-oracledb/issues/1400 going on.

**Comment by mike12345567** [comment]
> @mike12345567 Oh ok no worries, looks like I jumped the gun on this :) Thank you for all the work, super excited and look forward to the official ARM images once available.
> 
> If it's of any help, I looked into the build and node logs, the error is because ARM does not offer pre-built images for node-oracledb, this module is marked as optional, but for whatever reason the app node keeps restarting with this error. I found this discussion [oracle/node-oracledb#1400](https://github.com/oracle/node-oracledb/issues/1400) going on.

Hi @sudipmandal - thanks for looking into this, when I tried it on my Raspberry Pi I did indeed have to disable our Oracle integration to make it work - perhaps this is the way forward for the time being, disabling features that aren't easily made ARM compatible; we'll make some final decisions on this when we come round to actually finishing this issue off.

**Comment by mike12345567** [comment]
Closing in favour of #5202 which documents the requirement for ARM in the single docker image.
